sdv/sampling/hierarchical_sampler.py

```diff
@@ -113,32 +113,32 @@ def _add_child_rows(self, child_name, parent_name, parent_row, sampled_data):
                 sampled_data[child_name] = pd.concat(
                     [previous, sampled_rows]).reset_index(drop=True)
 
-    def _sample_table(self, synthesizer, table_name, num_rows, sampled_data):
-        """Sample a single table and all its children.
+    def _sample_table(self, table_name, sampled_data):
+        """Recursively sample every table.
+
+        Sample top level tables first, then their children, and so on.
 
         Args:
-            synthesizer (SingleTableSynthesizer):
-                Synthesizer to sample from for the table.
             table_name (string):
                 Name of the table to sample.
-            num_rows (int):
-                Number of rows to sample for the table.
             sampled_data (dict):
                 A dictionary mapping table names to sampled tables (pd.DataFrame).
         """
-        LOGGER.info(f'Sampling {num_rows} rows from table {table_name}')
+        for parent_name in self.metadata._get_parent_map()[table_name]:
+            if parent_name not in sampled_data:
+                self._sample_table(table_name=parent_name, sampled_data=sampled_data)
+                return  # Optimization to avoid iterating through the same nodes multiple times
```

Comment:
I think for HMA, it's simpler to skip this check and instead call `_sample_table` for each root node after we sample rows for it. If we have a child we've already sampled, we can skip sampling it (and any of its children) again.

Reply:
That runs into the problem I described in (1) [here](https://github.com/sdv-dev/SDV/pull/1562#issuecomment-1707217856), no?

Reply:
+1, for HMA we only use one parent to sample all the rows, then use likelihoods to add foreign keys for any other parents

Reply:
Theoretically only 1. The problem with this is, since we are doing DFS, all that really matters is the first root parent (depending on the graph ofc). So we'll first sample everything possible using the first root parent, and only then will we check if there are any extra parents. This is bad because (1) we ideally want all the parents to have influence over the graph and (2) the deeper the branch, the further we get from the original synthesizer in the root parent, so I'd imagine the quality of the results would deteriorate. I'm not sure how important this is in practice, but it's just something to keep in mind.

Because of that I tried the BFS approach (the approach (2) I explained): it's doing this same thing but actually going hierarchically, which minimizes both of those problems.

The other problem is that, if in the future we want to handle multiple foreign keys, this code would have to be completely rewritten, which is why I described approach (3).

Reply:
Why would the original approach not be able to handle foreign keys?

Reply:
You wouldn't have sampled the parents first. If you DFS from one root node, it's possible that the parents of the nodes you encounter along the way have not been sampled yet. So you need to do what I'm doing here, which is going up the graph whenever you encounter a non-sampled parent. And if we are going to be doing that, then there's no point in having that for loop over all the root nodes which we will find either way.

A question though, why would we choose approach (1) instead of (3)? From how I see it, (3) is just a slightly better version of (1).